### ISSUE TITLE
Removing undefined type from message fields

### ISSUE THREAD

**Issue by imirkin** [protobufjs/protobuf.js#1240]
Currently message fields are generated with the following type annotation:

    @member {foo.MessageFieldType|null|undefined} FieldName
    FooType.constructor.FieldName = null

Which in typescript translates into

    public FieldName?: (foo.MessageFieldType|null)

This means that the message class has optional fields, even though they're all set (to null) on the prototype. Obviously changing this would be a breaking change since people could have code that relies on the current types.

However, would it be reasonable to add an option, along the lines of --force-message (maybe --force-message-fields?), which removes the undefined bit?

I think this also matters when enabling `strictNullChecks` in the typescript compiler.

**Comment by NfNitLoop** [comment]
The [What's new in v6.0][1] wiki page says:

> There are no more null values marking unset fields. Instead, default values are on the prototype now and existence on the wire can be checked with Object#hasOwnProperty.

And I can see that behavior in the Javascript.  For example, I can see default values set on the JS prototype, like this:

```javascript
    /**
     * Post title.
     * @member {string} title
     * @memberof Post
     * @instance
     */
    Post.prototype.title = "";
```

So, not only can we remove the "undefined" but also the "null" in such cases.

However, `pbts` is still generating a .d.ts file that declares values may be undefined, like this:

```
    /** Post title */
    title?: (string|null);
```

[1]: https://github.com/protobufjs/protobuf.js/wiki/Changes-in-protobuf.js-6.0

**Comment by imirkin** [comment]
So my original request was probably somewhat misguided. I was complaining that sub-messages could be null. Which makes a ton of sense -- a non-null message is represented differently over the wire, even if it's otherwise empty.

The .d.ts file appears to have the null annotations for the I* types, but not for the full message types. I suspect this is intentional. I've avoided using the interface types in all but a handful of spots. Running pbjs with `--no-verify --no-convert --no-delimited --force-message --force-long` leads to pretty good type and code results, I've found. I was able to eventually enable `strictNullChecks` in a reasonable fashion throughout my (fairly large) codebase.

I'm going to close this (I forgot I had filed it!). @NfNitLoop if you feel like there's something wrong with the type annotations, I'd recommend you file your own issue detailing your use-case.
